cmd/utils/cmd.go

```diff
@@ -374,6 +374,85 @@ func ExportPreimages(db ethdb.Database, fn string) error {
 	return nil
 }
 
+// ExportSnapshotPreimages exports the preimages corresponding to the enumeration of
+// the snapshot for a given root.
+func ExportSnapshotPreimages(chain *core.BlockChain, fn string, root common.Hash) error {
+	log.Info("Exporting preimages", "file", fn)
+
+	fh, err := os.OpenFile(fn, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.ModePerm)
+	if err != nil {
+		return err
+	}
+	defer fh.Close()
+
+	writer := bufio.NewWriter(fh)
+	defer writer.Flush()
+
+	statedb, err := chain.State()
+	if err != nil {
+		return fmt.Errorf("failed to open statedb: %w", err)
+	}
+
+	if root == (common.Hash{}) {
+		root = chain.CurrentBlock().Root
+	}
+
+	type hashAndPreimageSize struct {
+		Hash common.Hash
+		Size int
+	}
+	hashCh := make(chan hashAndPreimageSize)
+
+	go func() {
+		defer close(hashCh)
+		accIt, err := chain.Snapshots().AccountIterator(root, common.Hash{})
+		if err != nil {
+			log.Error("failed to create account iterator", "error", err)
+			return
+		}
+		defer accIt.Release()
+
+		count := 0
+		for accIt.Next() {
+			acc, err := types.FullAccount(accIt.Account())
+			if err != nil {
+				log.Error("failed to get full account", "error", err)
+				return
+			}
+			hashCh <- hashAndPreimageSize{Hash: accIt.Hash(), Size: 20}
+
+			if acc.Root != (common.Hash{}) && acc.Root != types.EmptyRootHash {
+				stIt, err := chain.Snapshots().StorageIterator(root, accIt.Hash(), common.Hash{})
+				if err != nil {
+					log.Error("failed to create storage iterator", "error", err)
+					return
+				}
+				for stIt.Next() {
+					hashCh <- hashAndPreimageSize{Hash: stIt.Hash(), Size: 32}
+				}
+				stIt.Release()
+			}
+			count++
+			if count%100000 == 0 {
+				log.Info("Last exported account", "account", accIt.Hash())
+			}
+		}
+	}()
+
+	for item := range hashCh {
+		preimage := rawdb.ReadPreimage(statedb.Database().DiskDB(), item.Hash)
+		if len(preimage) != item.Size {
+			return fmt.Errorf("invalid preimage size")
+		}
```

Comment:
When we're missing a preimage, now it would just die with `invalid preimage size'. Wouldn't it be nicer to spit out at least the missing preimage (and maybe even continue: printing out all missing preimages, so the user can look them up and then redo the conversion?)